File: JavaScript/17_Binary Tree-I/E145. Postorder Traversal.js
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/

/**
 * TreeNode class for binary tree
 */
/**
 * Creates a tree node
 */
function createTreeNode(val = 0, left = null, right = null) {
    return { val, left, right };
}

/**
 * Iterative Postorder Traversal using stack (reverse preorder trick)
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number[]} - Postorder traversal (Left -> Right -> Root)
 */
function postorderTraversal(root) {
    const result = [];
    if (root === null) return result;

    const stack = [root];

    while (stack.length > 0) {
        const node = stack.pop();
        result.push(node.val);

        // Push left first so right is processed first
        if (node.left !== null) stack.push(node.left);
        if (node.right !== null) stack.push(node.right);
    }

    // Reverse to get postorder
    return result.reverse();
}

/**
 * Recursive Postorder Traversal
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number[]} - Postorder traversal
 */
function postorderTraversalRecursive(root) {
    const result = [];

    function postorder(node) {
        if (node === null) return;
        postorder(node.left);
        postorder(node.right);
        result.push(node.val);
    }

    postorder(root);
    return result;
}

// Test cases
const root = createTreeNode(1);
root.left = createTreeNode(2);
root.right = createTreeNode(3);
root.left.left = createTreeNode(4);
root.left.right = createTreeNode(5);
root.right.left = createTreeNode(6);
root.right.right = createTreeNode(7);

console.log(postorderTraversal(root)); // [4, 5, 2, 6, 7, 3, 1]
console.log(postorderTraversalRecursive(root)); // [4, 5, 2, 6, 7, 3, 1]

File: Java/17_Binary Tree-I/E145. Postorder Traversal.java
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

class Solution{
     public static class TreeNode {
          int val;
          TreeNode left;
          TreeNode right;
          TreeNode() {}
          TreeNode(int val) { this.val = val; }
          TreeNode(int val, TreeNode left, TreeNode right) {
              this.val = val;
              this.left = left;
              this.right = right;
          }
      }
      
      public List<Integer> postorderTraversal(TreeNode root) {
              
                List<Integer> result = new ArrayList<>();
                if (root == null) return result;
                
                Stack<TreeNode> stack = new Stack<>();
                stack.push(root);
                
                while (!stack.isEmpty()) {
                    TreeNode node = stack.pop();
                    result.add(node.val);
                    
                    if (node.left != null) stack.push(node.left);
                    if (node.right != null) stack.push(node.right);
                }
                
                Collections.reverse(result);
                return result;
      }
      
      public static void main(String[] args) {
          Solution solution = new Solution();
          TreeNode root = new TreeNode(1);
          root.left = new TreeNode(2);
          root.right = new TreeNode(3);
          root.left.left = new TreeNode(4);
          root.left.right = new TreeNode(5);
          root.right.left = new TreeNode(6);
          root.right.right = new TreeNode(7);
          
          List<Integer> result = solution.postorderTraversal(root);
          System.out.println(result);
      }
}